lib/decode/decode.go

```diff
@@ -0,0 +1,113 @@
+/*Package decode provides tools for customizing the decoding of configuration
+blogs, into structures using mapstructure.
+*/
+package decode
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// KeyTranslator identifies a structure as one which should have its keys
+// translated by HookTranslateKeys.
+type KeyTranslator interface {
+	// DecodeKeyMapping must return a mapping of:
+	//     Lower case deprecated key -> canonical key
+	// If the data contains a key which matches the deprecated key, the value
+	// for that key is moved to the canonical key. The deprecated key is matched
+	// case-insensitive.
+	// If the canonical key already exists in the data, the deprecated key is
+	// not modified.
+	DecodeKeyMapping() map[string]string
+}
+
+var typeOfKeyTranslator = reflect.TypeOf((*KeyTranslator)(nil)).Elem()
+
+// HookTranslateKeys is a mapstructure decode hook which translates keys in a
+// map to their canonical value. Implement KeyTranslator on a struct to
+// have HookTranslateKeys translate keys. See KeyTranslator for more details.
+//
+// This hook should ONLY be used to maintain backwards compatibility with
+// deprecated keys. For new structures use struct tags to set the desired
+// serialization key.
+//
+// TODO: If field aliases were identified as struct tags, instead of a map
+// returned from a method on the struct, we could avoid creating instances of
+// the type.
+func HookTranslateKeys(_, to reflect.Type, data interface{}) (interface{}, error) {
+	// When the target is a pointer mapstructure will call the hook again with
+	// the struct. Return immediately if target is a pointer to avoid doing the
+	// work twice.
+	if to.Kind() == reflect.Ptr {
+		return data, nil
+	}
+
+	// Avoid doing any work if data is not a map
+	target, ok := data.(map[string]interface{})
+	if !ok {
+		return data, nil
+	}
+
+	var translator KeyTranslator
+	switch {
+	case to.Implements(typeOfKeyTranslator):
+		translator = reflect.Zero(to).Interface().(KeyTranslator)
+	case reflect.PtrTo(to).Implements(typeOfKeyTranslator):
+		translator = reflect.New(to).Interface().(KeyTranslator)
+	default:
+		return data, nil
+	}
```

Comment:
You might be able to replace all of this with something like:

```
val := to.Indirect()
if !val.Implements(typeOfKeyTranslator) {
   return data, nil
}

translator := val.Interface().(KeyTranslator)
```

Reply:
Its not much shorter but I would find it easier to read.

Reply:
`to` is a `reflect.Type` not a `reflect.Value`. `reflect.Indirect` works on a `Value`.

I started with something similar to what you suggest, but it failed some of my test cases. There are 4 possible permutations which need to be supported:

1. `to` is a pointer, and `DecodeKeyMapping` has a pointer receiver
2. `to` is a pointer, and `DecodeKeyMapping` has a value receiver
3. `to` is a struct, and `DecodeKeyMapping` has a pointer receiver
4. `to` is a struct, and `DecodeKeyMapping` has a value receiver

In regular code the 2nd case is automatically handled by the compiler, but I don't think that works when using `reflect.Value`.

To handle these permutations, the first guard (`to.Kind() == reflect.Ptr`) handles case 1 and 2 by doing nothing, relying on mapstructure to call the hook again with the struct.
The remaining two cases are handled in this switch/case. I believe both cases are necessary because the interface may have been implemented on either the struct or the pointer type, and we have to create the appropriate type. We need the opposite of `Indirect()` , which I believe is `reflect.New()`.

It's possible there is a way to avoid this extra case, but I have not found it. I will add some more documentation to try and make this more clear.

Reply:
I'm also hoping this is only a transitional fix. That we will be able to replace this hook with one that uses struct field tags instead of a method on the struct. I think that might be a larger change, so I haven't tried it yet.
If this hook isn't obvious maybe we can try that approach instead of this one.

Reply:
I tried the struct tag approach, and it looks much better. I left the old approach in a separate commit.